kong/degraphql_route_service_test.go

```diff
@@ -0,0 +1,289 @@
+	require.Equal(t, "foo2", *serviceB.Name)
+	t.Cleanup(func() {
+		require.NoError(t, client.Services.Delete(defaultCtx, serviceB.ID))
+	})
+
+	t.Run("add routes to a service, list them", func(t *testing.T) {
+		routeTemplates := []*DegraphqlRoute{
+			{
+				Service: serviceA,
+				URI:     String("/front"),
+				Query:   String("query { viewer { login } }"),
+			},
+			{
+				Service: serviceA,
+				URI:     String("/frontBis"),
+				Query:   String("query { viewer { login } }"),
+			},
+			{
+				Service: serviceA,
+				URI:     String("/frontThird"),
+				Query:   String("query { viewer { login } }"),
+			},
+			{
+				Service: serviceA,
+				URI:     String("/front4"),
+				Query:   String("query { viewer { login } }"),
+			},
+			{
+				Service: serviceA,
+				URI:     String("/frontLast"),
+				Query:   String("query { viewer { login } }"),
+			},
+		}
+
+		for _, r := range routeTemplates {
+			createdRoute, err := client.DegraphqlRoutes.Create(defaultCtx, r)
+			require.NoError(t, err)
+			require.NotNil(t, createdRoute)
+			require.NotEmpty(t, *createdRoute.ID)
+			require.NotEmpty(t, createdRoute.Service.Name)
+
+			*r = *createdRoute
+		}
+		t.Cleanup(func() {
+			for _, r := range routeTemplates {
+				require.NoError(t, client.DegraphqlRoutes.Delete(defaultCtx, r.Service.ID, r.ID))
+			}
+		})
+
+		sort.Sort(ByID(routeTemplates))
+
+		routes, next, err := client.DegraphqlRoutes.List(defaultCtx, serviceA.Name, &ListOpt{})
+		require.NoError(t, err)
+		require.Nil(t, next)
+
+		sort.Sort(ByID(routes))
+		for i, r := range routeTemplates {
+			require.Equal(t, *r.ID, *routes[i].ID)
+			require.Equal(t, *r.Service.ID, *routes[i].Service.ID)
+			require.Equal(t, *r.URI, *routes[i].URI)
+			require.Equal(t, *r.Query, *routes[i].Query)
+		}
+	})
+
+	t.Run("add routes to different service, list each", func(t *testing.T) {
+		routeTemplates := []*DegraphqlRoute{
+			{
+				Service: serviceA,
+				URI:     String("/front"),
+				Query:   String("query { viewer { login } }"),
+			},
+			{
+				Service: serviceB,
+				URI:     String("/frontBis"),
+				Query:   String("query { viewer { login } }"),
+			},
+			{
+				Service: serviceB,
+				URI:     String("/frontThird"),
+				Query:   String("query { viewer { login } }"),
+			},
+			{
+				Service: serviceA,
+				URI:     String("/front4"),
+				Query:   String("query { viewer { login } }"),
+			},
+			{
+				Service: serviceA,
+				URI:     String("/frontLast"),
+				Query:   String("query { viewer { login } }"),
+			},
+		}
+
+		for _, r := range routeTemplates {
+			createdRoute, err := client.DegraphqlRoutes.Create(defaultCtx, r)
+			require.NoError(t, err)
+			require.NotNil(t, createdRoute)
+			require.NotEmpty(t, *createdRoute.ID)
+			require.NotEmpty(t, createdRoute.Service.Name)
+
+			*r = *createdRoute
+		}
+		t.Cleanup(func() {
+			for _, r := range routeTemplates {
+				require.NoError(t, client.DegraphqlRoutes.Delete(defaultCtx, r.Service.ID, r.ID))
+			}
+		})
```

Comment:
To consider:

```suggestion
		for _, r := range routeTemplates {
			createdRoute, err := client.DegraphqlRoutes.Create(defaultCtx, r)
			require.NoError(t, err)
			require.NotNil(t, createdRoute)
			t.Cleanup(func() {
				serviceID := r.Service.ID
				routeID := r.ID
				require.NoError(t, client.DegraphqlRoutes.Delete(defaultCtx, serviceID, routeID))
			})
			require.NotEmpty(t, *createdRoute.ID)
			require.NotEmpty(t, createdRoute.Service.Name)

			*r = *createdRoute
		}
```

Reply:
I did that first, and then lost many hours debugging until I realised the pointer references didn't work well from within the loop.  that's why i moved the loop to within the cleanup callback and it solved it all.

tried it again just now, hoping the local copies of `serviceID` and `routeID` would help; but fails in the same way.  I guess it's because those variables are actually pointers to strings, not strings themselves and the extra level of indirection gets in the way.

looks nicer, but it's too fragile to be worth it.

Reply:
Oh, my bad I didn't notice those are pointers.

In that case what you can do is this:

```suggestion
		for _, r := range routeTemplates {
			createdRoute, err := client.DegraphqlRoutes.Create(defaultCtx, r)
			require.NoError(t, err)
			require.NotNil(t, createdRoute)
			require.NotEmpty(t, *createdRoute.ID)

			serviceID := *r.Service.ID
			routeID := *createdRoute.ID
			t.Cleanup(func() {
				assert.NoError(t, client.DegraphqlRoutes.Delete(defaultCtx, &serviceID, &routeID))
			})
			require.NotEmpty(t, createdRoute.Service.Name)

			*r = *createdRoute
		}
```

I've just run it locally and it worked just fine for me (after setting up kong ee via `setup_kong_ee.sh` which has some problems but that's another topic).

BTW: this `*r = *createdRoute` is a code smell to me. I'd much rather see this placed in a separate slice that's dedicated to store created routes.

Reply:
Leaving this as food for thought for later. No need to implement it here.